internal/configs/test_provider_body.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) The OpenTofu Authors
+// SPDX-License-Identifier: MPL-2.0
+// Copyright (c) 2023 HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package configs
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+var _ hcl.Body = &testProviderBody{}
+
+// testProviderBody is a wrapper around hcl.Body that allows us to use prepared EvalContext for provider configuration.
+type testProviderBody struct {
+	originalBody hcl.Body
+	evalCtx      *hcl.EvalContext
+}
+
+func (c testProviderBody) Content(schema *hcl.BodySchema) (*hcl.BodyContent, hcl.Diagnostics) {
+	content, diags := c.originalBody.Content(schema)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+	attrs := content.Attributes
+	for name, attr := range attrs {
```

Comment:
Looks like this shares a lot of code between Content and PartialContent.  Perhaps a common function for transforming hcl.BodyContent would help.